librato_python_web/instrumentor/external/urllib2_.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) 2015. Librato, Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#     * Redistributions of source code must retain the above copyright
+#       notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#     * Neither the name of Librato, Inc. nor the names of project contributors
+#       may be used to endorse or promote products derived from this software
+#       without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL LIBRATO, INC. BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+from math import floor
+
+from librato_python_web.instrumentor import context
+from librato_python_web.instrumentor.instrument import function_wrapper_factory, _should_be_instrumented
+from librato_python_web.instrumentor.base_instrumentor import BaseInstrumentor
+from librato_python_web.instrumentor import telemetry
+from librato_python_web.instrumentor.util import get_parameter, Timing
+
+
+# TODO: make this generally available in instrument.py??
+def _wrapped_call(metric, func, *args, **keywords):
+    """ Times and executes arbitrary method """
+    state = 'external'
+
+    if not _should_be_instrumented(state, enable_if='web', disable_if='model'):
+        return func(*args, **keywords)
+
+    Timing.push_timer()
+    try:
+        context.push_state(state)
+        return func(*args, **keywords)
+    finally:
+        context.pop_state(state)
+        elapsed, _ = Timing.pop_timer()
+        telemetry.record(metric, elapsed)
```

Comment:
Nice! `state` obviously needs to be made into a parameter or a factory wrapper, but that looks reusable.